docs/auth-engines.md

```diff
@@ -92,3 +93,104 @@ This CR is roughly equivalent to this Vault CLI command:
 ```shell
 vault write [namespace/]auth/kubernetes/role/database-engine-admin bound_service_account_names=vaultsa bound_service_account_namespaces=<dynamically generated> policies=database-engine-admin
 ```
+
+## LDAPAuthEngineConfig
+
+The `LDAPAuthEngineConfig` CRD allows a user to configure an authentication engine mount of [type LDAP](https://www.vaultproject.io/docs/auth/ldap).
+
+```yaml
+apiVersion: redhatcop.redhat.io/v1alpha1
+kind: LDAPAuthEngineConfig
+metadata:
+  name: authenginemount-sample
+spec:
+  UPNDomain:
+  anonymousGroupSearch:
+  authentication: 
+    path: kubernetes
+    role: policy-admin
+    serviceAccount:
+      name: admin-sa
+  bindDN: cn=vault,ou=Users,dc=example,dc=com
+  bindPass: xxxxxxxxxxxxx
```

Comment:
this should be passed as a secret, please refer to other engines that do similar things

Reply:
I try to do it same way as DatabaseSecretEngineConfig type. CRDs are generated correctly, but the bindDN & bindPass fields somehow do not get populated : 

```
  bindCredentials:
    passwordKey: password
    secret:
      name: bindcredentials
    usernameKey: username
  bindDN: ""
  bindPass: ""
```

Reply:
you have to initialize those fields in the `PrepareInternalValues` fuinction. `bindPass`, should not be exposed by the CRD.
Let me know if you need more help.
see also: https://github.com/redhat-cop/vault-config-operator/blob/7edf4ffb162efa3d4232c98c604146f40bf839d2/api/v1alpha1/databasesecretengineconfig_types.go#L73

Reply:
Hi @raffaelespazzoli thank you for the feedback.
tbh I followed the same approach as _DatabaseSecretEngineConfig_ type and set both methods _PrepareInternalValues_ & _setInternalCredentials_ accordingly. Did a few more changes, but still I am not able to retrieve those values from _bindCredentials_ property.

Reply:
I am not sure what you mean by "retrieving those credentials". You really are not supposed to retrieve them. Can you explain what you are trying to do? Vault will not show them back to you.

Reply:
Hi @raffaelespazzoli, I thought that when creating the CR for _LDAPAuthEngineConfig_, the values for _bindDN_ & _bindPass_ would get populated by _bindCredentials_, but instead I get : 

provided CR : 

```
  url: "ldaps://ldap-xxx-a.xxx.xxx.xxx,ldaps://ldap-xxx-b.xxx.xxx.xxx,ldaps://ldap-xxx-c.xxx.xxx.xxx"
  caseSensitiveNames: false
  bindDN: "supersecret"
  bindCredentials:
    secret: 
      name: bindcredentials
    usernameKey: username
    passwordKey: password
  userDN: "ou=Users,ou=XXXX,ou=Company,dc=xxx,dc=xxx,dc=xxx" 
```

populated CR (after creation): 

```
    authentication:
      path: kubernetes
      role: policy-admin
      serviceAccount:
        name: default
    bindCredentials:
      passwordKey: password
      secret:
        name: bindcredentials
      usernameKey: username
    bindDN: supersecret
    bindPass: ""
    caseSensitiveNames: false
```
Since these two properties (_bindDN_ & _bindPass_) are also part of the payload send to Vault API, I though these should be somehow visible.

Reply:
mm, I'm lost...I don't understand what you are trying to do, the "supersecret" value in bindDN is confusing. Anyway, `bindPass` is clearly a secret and should NOT appear in the CR. Looking at your code, it seems to be the case. I don't know LDAP in detail, if `bindDN` is also a secret, you should treat it the same way as `bindPass`. If not, what you have in the first code quote seems correct.
I don't understand what the second quote is supposed to be.., is that what you would expect to see?

Reply:
Hi @raffaelespazzoli , sorry for the confusion. The bindDN field could or not be a secret. This is optional and up to the user. Regarding the bindPass, you are correct, it shouldn't appear in the CR. 
Exactly, the second quote is what I thought would be the correct one, but it was not the case, as you pointed out.
Do you have any suggestion how I can test this ? Should I spin up a Minishift/CodeReady/kind cluster + OpenLDAP ? I can't test it at the moment on a OpenShift Prod Cluster.
thnx

Reply:
Yes, we try to test our stuff with a real integration test before releasing it. At a minimum script the test so it can be execute manually. Better would be to automate it, given that now we can do it with this operator.
A while ago a scripted how to deploy an LDAP server on OCP and populate with some test data. Maybe you can start from here: https://cloud.redhat.com/blog/orgs-management-and-team-onboarding-in-openshift-a-fully-automated-approach.

Reply:
Hi @raffaelespazzoli, 
I took some time to study your repo for `orgs-management-ocp/team-onboarding` topic you provided. I am still working to automate the integration test. Meanwhile, I am doing some manual tests : 

1. Spin-up a `kind `cluster where both `Vault ` (`dev` mode) and `LDAP ` are deployed (they can see each other in terms of network)
2. Create an `AuthEngineMount `CR named  `ldap-test`
3. Create an `LDAPAuthEngineConfig` CR as follows : 

```
  path: ldap-test
  url: "ldap://ldap.ldap.svc.cluster.local"
  caseSensitiveNames: false
  bindDN: "cn=admin,cn=config"
  userDN: "dc=example,dc=com"
  bindCredentials:
    secret: 
      name: bindcredentials
    usernameKey: username
    passwordKey: password
  insecureTls: true  ---> TLS has been disabled also in LDAP side
```
It seems that at least the `bindCredentials` is working okay, since I can login there, but I get the following errors : 

`2022-04-29T15:01:30.536Z [DEBUG] auth.ldap.auth_ldap_9953873a: error getting user bind DN: error="LDAP search for binddn failed: LDAP Result Code 32 \"No Such Object\": "`

if that was not the case, we would get : 

`022-04-29T14:03:09.371Z [DEBUG] auth.ldap.auth_ldap_9953873a: error getting user bind DN: error="LDAP search for binddn failed: LDAP bind (service) failed: LDAP Result Code 49 \"Invalid Credentials\": "`

Since I don't know LDAP in detail, I am not sure what I am missing  in the LDAP config (CR).
I am using the LDAP test data part of your repo. Login in Vault (via ldap) :  `username`: trevor , `password `:  xxxxx

Reply:
UPDATE: 
Apparently, I was using the wrong `bindDN `& `baseDN`.
I used the following config (CR) instead : 

```
  path: ldap-test
  url: "ldap://ldap.ldap.svc.cluster.local"
  caseSensitiveNames: false
  bindDN: "cn=admin,dc=example,dc=com"
  userDN: "ou=users,dc=example,dc=com"
  bindCredentials:
    secret: 
      name: bindcredentials
    usernameKey: username
    passwordKey: password
  insecureTls: true
```

and I am able to login with username/uid : `trevor `and its password.

Vault & LDAP logs : 

Vault 

```
2022-05-02T21:06:25.566Z [DEBUG] auth.ldap.auth_ldap_ee620879: compiling search filter: search_filter="({{.UserAttr}}={{.Username}})"
2022-05-02T21:06:25.566Z [DEBUG] auth.ldap.auth_ldap_ee620879: discovering user: userdn="ou=users,dc=example,dc=com" filter="(cn=trevor)"
2022-05-02T21:06:25.567Z [DEBUG] auth.ldap.auth_ldap_ee620879: user binddn fetched: username=trevor binddn="uid=trevor,ou=Users,dc=example,dc=com"
2022-05-02T21:06:25.567Z [DEBUG] auth.ldap.auth_ldap_ee620879: re-bound to original binddn
2022-05-02T21:06:25.567Z [WARN]  auth.ldap.auth_ldap_ee620879: groupfilter is empty, will not query server
2022-05-02T21:06:25.567Z [DEBUG] auth.ldap.auth_ldap_ee620879: groups fetched from server: num_server_groups=0 server_groups=[]
2022-05-02T21:06:25.567Z [DEBUG] auth.ldap.auth_ldap_ee620879: compiling search filter: search_filter="({{.UserAttr}}={{.Username}})"
2022-05-02T21:06:25.568Z [DEBUG] auth.ldap.auth_ldap_ee620879: discovering user: userdn="ou=users,dc=example,dc=com" filter="(cn=trevor)"
```

LDAP

```
62704834 conn=1030 fd=12 ACCEPT from IP=10.42.0.178:56588 (IP=0.0.0.0:389)
62704834 conn=1030 op=0 BIND dn="cn=admin,dc=example,dc=com" method=128
62704834 conn=1030 op=0 BIND dn="cn=admin,dc=example,dc=com" mech=SIMPLE ssf=0
62704834 conn=1030 op=0 RESULT tag=97 err=0 text=
62704834 conn=1030 op=1 SRCH base="ou=users,dc=example,dc=com" scope=2 deref=0 filter="(cn=trevor)"
62704834 conn=1030 op=1 SRCH attr=cn
62704834 <= mdb_equality_candidates: (cn) not indexed
62704834 conn=1030 op=1 SEARCH RESULT tag=101 err=0 nentries=1 text=
62704834 conn=1030 op=2 BIND anonymous mech=implicit ssf=0
62704834 conn=1030 op=2 BIND dn="uid=trevor,ou=Users,dc=example,dc=com" method=128
62704834 conn=1030 op=2 BIND dn="uid=trevor,ou=Users,dc=example,dc=com" mech=SIMPLE ssf=0
62704834 conn=1030 op=2 RESULT tag=97 err=0 text=
62704834 conn=1030 op=3 BIND anonymous mech=implicit ssf=0
62704834 conn=1030 op=3 BIND dn="cn=admin,dc=example,dc=com" method=128
62704834 conn=1030 op=3 BIND dn="cn=admin,dc=example,dc=com" mech=SIMPLE ssf=0
62704834 conn=1030 op=3 RESULT tag=97 err=0 text=
62704834 conn=1030 op=4 BIND anonymous mech=implicit ssf=0
62704834 conn=1030 op=4 BIND dn="cn=admin,dc=example,dc=com" method=128
62704834 conn=1030 op=4 BIND dn="cn=admin,dc=example,dc=com" mech=SIMPLE ssf=0
62704834 conn=1030 op=4 RESULT tag=97 err=0 text=
62704834 conn=1030 op=5 SRCH base="ou=users,dc=example,dc=com" scope=2 deref=0 filter="(cn=trevor)"
62704834 conn=1030 op=5 SRCH attr=cn
62704834 <= mdb_equality_candidates: (cn) not indexed
62704834 conn=1030 op=5 SEARCH RESULT tag=101 err=0 nentries=1 text=
62704834 conn=1030 fd=12 closed (connection lost)
```

Do you think that this is sufficient for the integration test ?
Thnx

Reply:
yes I think this is good enough

Reply:
Do I need to do anymore here, before we set this to "Resolved" ?